Userland/Libraries/LibJS/Runtime/Completion.cpp

```diff
@@ -89,21 +90,27 @@ ThrowCompletionOr<Value> await(GlobalObject& global_object, Value value)
     auto on_rejected = NativeFunction::create(global_object, "", move(rejected_closure));
 
     // 7. Perform ! PerformPromiseThen(promise, onFulfilled, onRejected).
-    verify_cast<Promise>(promise)->perform_then(on_fulfilled, on_rejected, {});
+    auto* promise = verify_cast<Promise>(promise_object);
+    promise->perform_then(on_fulfilled, on_rejected, {});
+
+    // FIXME: Since we don't support context suspension, we attempt to "wait" for the promise to resolve
+    //        by letting the event loop spin until our promise is no longer pending, and then synchronously
+    //        running all queued promise jobs.
+    if (Core::EventLoop::has_been_instantiated())
+        Core::EventLoop::current().spin_until([&] { return promise->state() != Promise::State::Pending; });
```

Comment:
we should probably `vm.run_queued_promise_jobs();` before hand, or else this would spin forever.

Reply:
How so? we have to first wait for the promise to resolve, _then_ run the promise jobs (otherwise we'd just end up never calling either of the resolve or reject functions...)

Reply:
The promise state will change by an outside source (e.g. `XMLHttpRequest`'s `load` event firing and calling `resolve()`), to which we might not _react_ until `vm.run_queued_promise_jobs()` is run.

Reply:
but for the promise state to change by XHR, we first have to call XHR to start the request, which will only happen once we `vm.run_queued_promise_jobs()`....

Reply:
Wait, why?
XHR has already started at this point, we ran the whole `new Promise(...)` thing as part of evaluating the awaited expression, so now we just have to wait for it to finish and call our `resolve` (or `reject`), that has nothing to do with promise jobs

Reply:
by calling `new Promise()` we don't run the given promise callback, we simply register it as a queued promise job.

Reply:
Ohhh I see, that only happens if the callback is an async function, so in this case it is should be synchronous, my bad

Reply:
That doesn't sound right...

Reply:
I mean it's what we do, we immediately call the promise callback in the promise constructor, so unless it's async (which just queues another promise) it will run synchronously.

Reply:
we definitely do call the "executor":
https://github.com/SerenityOS/serenity/blob/58c6a156bf3e6dcf0e37a3a73ba6dd4365ce1203/Userland/Libraries/LibJS/Runtime/PromiseConstructor.cpp#L347

Reply:
> if the callback is an async function

That sounds like a big mess, is that actually supported?

Reply:
Right, that's what I said

Reply:
> > if the callback is an async function
> 
> That sounds like a big mess, is that actually supported?

`(new Promise(async (resolve) => { return resolve(3); })).then(console.log);`

Reply:
Looks like the spec just doesn't care whether it's async or not, makes sense - it'll resolve _eventually_ if it's async I guess.

I'm pretty sure this impl would be fine with that, since we're serialising `await`, the "executor" will block in the Promise constructor.

Reply:
I'm not 100% sure I understand the flow correctly now, but that's what test262 tests are for, as long as this doesn't regress any :+1:  from me